streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -982,17 +1000,34 @@ private synchronized void setStateWhenNotInPendingShutdown(final State newState)
         setState(newState);
     }
 
-    private synchronized void setState(final State newState) {
-        final State oldState = state;
-        if (!state.isValidTransition(newState)) {
-            log.warn("{} Unexpected state transition from {} to {}.", logPrefix, oldState, newState);
-        } else {
-            log.info("{} State transition from {} to {}.", logPrefix, oldState, newState);
-        }
+    /**
+     * Sets the state
+     * @param newState New state
+     */
+    void setState(final State newState) {
+        synchronized (stateLock) {
+            final State oldState = state;
+
+            // there are cases when we shouldn't check if a transition is valid, e.g.,
+            // when, for testing, a thread is closed multiple times. We could either
+            // check here and immediately return for those cases, or add them to the transition
+            // diagram (but then the diagram would be confusing and have transitions like
+            // PENDING_SHUTDOWN->PENDING_SHUTDOWN).
+            if (newState != State.DEAD && (state == State.PENDING_SHUTDOWN || state == State.DEAD)) {
```

Comment:
@enothereska As I mentioned above, could you list all the transitions that we are trying to bypass with this condition so we can discuss each case? Similarly for `KafkaStreams.setState`'s if condition to bypass the validation, could you also list all the transitions that we should be targeting for?